OSOE-353: Add duplicate SQL query detector in Lombiq.UITestingToolbox

dministro opened the pull request:
[OSOE-353](https://lombiq.atlassian.net/browse/OSOE-353)
Fixes #93 

[OSOE-353]: https://lombiq.atlassian.net/browse/OSOE-353?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
  - Introduced a "Duplicated SQL query detector" to improve test efficiency and resource usage.
  - Added configuration settings for UI Testing Toolbox to enhance SQL query monitoring.
  - Implemented a new set of tests for detecting duplicated SQL queries.

- **Enhancements**
  - Improved test environment with additional configuration options for database command execution thresholds.
  - Enhanced navigation and page load testing with new counter data collection and assertion capabilities.
  - Expanded testing toolbox with classes and interfaces to support performance metric collection.

- **Documentation**
  - Updated the Readme to reflect new testing features and configurations.

- **Bug Fixes**
  - Fixed issues related to database command execution tracking during tests.

- **Refactor**
  - Refactored database connection and command wrappers to support performance data collection.

- **Tests**
  - Added tests for new functionality and performance monitoring features.

- **Chores**
  - Updated internal test infrastructure to incorporate new counter data collector services.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dministro committed:
Adding counter infrastructure

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

0liver commented on Lombiq.Tests.UI/Exceptions/CounterThresholdException.cs:
```diff
@@ -0,0 +1,50 @@
+
+namespace Lombiq.Tests.UI.Exceptions;
+
+// We need constructors with required informations.
+#pragma warning disable CA1032 // Implement standard exception constructors
```
The warning does not disallow custom constructors, but it enforces implementation of all default constructors incl. one to support serialization.

I don't think we should opt out of that.

sarahelsaig replied:
I concur, you must have [these constructors](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1032#rule-description) for safe error handling, but you can have additional constructors if you want to include custom logic.

dministro committed:
Using interfaces

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Implementing required exception constructors

dministro committed:
Adjusting tests
Fixing typos

dministro committed:
Fixing typo

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

Piedone commented on Lombiq.Tests.UI/Services/PhaseCounterConfiguration.cs:
```diff
@@ -0,0 +1,13 @@
+    public int DbCommandExecutionRepetitionPerNavigationThreshold { get; set; } = 11;
+    public int DbCommandExecutionRepetitionThreshold { get; set; } = 22;
+    public int DbReaderReadPerNavigationThreshold { get; set; } = 11;
+    public int DbReaderReadThreshold { get; set; } = 11;
```
Add docs to the properties.

Piedone commented on Lombiq.Tests.UI/Services/PhaseCounterConfiguration.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Lombiq.Tests.UI.Services;
+
+public class PhaseCounterConfiguration
```
Mention this feature in the root Readme.

Piedone replied:
This still needs to be done.

Piedone commented on Lombiq.Tests.UI/Services/UITestContext.cs:
```diff
@@ -100,13 +100,19 @@ public class UITestContext
+    /// <summary>
+    /// Gets the currently running <see cref="CounterDataCollector"/> instance.
+    /// </summary>
+    public CounterDataCollector CounterDataCollector { get; init; }
```
This needs to be possible to be completely disabled too.

Piedone replied:
This still needs to be done. The page change dumps should also only be written when duplicate SQL query detection is enabled.

Piedone commented on Lombiq.Tests.UI/Services/CounterDataCollector.cs:
```diff
@@ -0,0 +1,34 @@
+        _probes.SelectWhere(probe => probe, probe => probe.IsRunning)
+            .ForEach(probe => probe.Increment(counter));
+
+        base.Increment(counter);
```
```suggestion
        _probes.Where(probe => probe.IsRunning).ForEach(probe => probe.Increment(counter));
        base. Increment(counter);
```

dministro replied:
The `ConcurrentBag` doesn't have `Where` method.

Piedone replied:
There's an extension for it in `System.Linq` as usual. Do check though if such iterating needs any special care due to it being a concurrent collection.

(BTW `SelectWhere` is form Helpful Libraries.)

Piedone commented on Lombiq.Tests.UI/Services/Counters/CounterProbeBase.cs:
```diff
@@ -0,0 +1,39 @@
+        if (counter is not TCounter value)
+        {
+            throw new ArgumentException(
+                $"The type of ${nameof(counter)} is not compatible with ${typeof(TCounter).Name}");
```
```suggestion
                $"The type of ${nameof(counter)} is not compatible with ${typeof(TCounter).Name}.");
```

Piedone commented on Lombiq.Tests.UI/Services/Counters/ICounterProbe.cs:
```diff
@@ -0,0 +1,38 @@
+namespace Lombiq.Tests.UI.Services.Counters;
+
+/// <summary>
+/// Represents a probe for the counter infrastructure.
```
Explain what a probe is: "Represents a probe (...)..."

Piedone commented on Lombiq.Tests.UI/Services/Counters/ICounterValue.cs:
```diff
@@ -0,0 +1,13 @@
+    /// <summary>
+    /// Dumps the value content to a human readable format.
+    /// </summary>
+    /// <returns>A human readable string representation of instance.</returns>
```
```suggestion
    /// <returns>A human-readable string representation of the instance.</returns>
```

Piedone commented on Lombiq.Tests.UI/Services/Counters/ICounterProbe.cs:
```diff
@@ -0,0 +1,38 @@
+    void Increment(ICounterKey counter);
+
+    /// <summary>
+    /// Dumps the probe headline to a human readable format.
```
"human-readable" everywhere.

Piedone commented on Lombiq.Tests.UI/Services/Counters/NavigationProbe.cs:
```diff
@@ -0,0 +1,14 @@
+        : base(counterDataCollector) =>
+        AbsoluteUri = absoluteUri;
+
+    public override string DumpHeadline() => $"{nameof(NavigationProbe)}, AbsoluteUri={AbsoluteUri}";
```
```suggestion
    public override string DumpHeadline() => $"{nameof(NavigationProbe)}, AbsoluteUri = {AbsoluteUri}";
```

Piedone commented on Lombiq.Tests.UI/Extensions/NavigationUITestContextExtensions.cs:
```diff
@@ -39,7 +40,10 @@ public static class NavigationUITestContextExtensions
+                using (new NavigationProbe(context.CounterDataCollector, absoluteUri))
+                {
+                    context.Driver.Navigate().GoToUrl(absoluteUri);
+                }
```
Note that this is not the only way we navigate between pages. Check `UITestExecutionEvents` for safer ways to make sure you run something before and after navigation.

Piedone commented on Lombiq.Tests.UI.Samples/Tests/BasicOrchardFeaturesTests.cs:
```diff
@@ -25,7 +25,17 @@ public BasicOrchardFeaturesTests(ITestOutputHelper testOutputHelper)
+                // The UI Testing Toolbox includes DbCommand execution counter, after the end of test it checks the
+                // number of executed commands with the same Sql command and parameter set against the threshold value
+                // in its configuration. If the executed command count is greater then the threshold, it raises a
+                // CounterThresholdException. So here we set the minimum required value to avoid it.
```
```suggestion
                // The UI Testing Toolbox includes a DbCommand execution counter. After the end of the test, it checks the
                // number of executed commands with the same SQL command and parameter set against the threshold value
                // in its configuration. If the executed command count is greater than the threshold, it raises a
                // CounterThresholdException. So here we set the minimum required value to avoid it.
```

Piedone commented on Lombiq.Tests.UI.Samples/Tests/BasicOrchardFeaturesTests.cs:
```diff
@@ -25,7 +25,17 @@ public BasicOrchardFeaturesTests(ITestOutputHelper testOutputHelper)
+                // number of executed commands with the same Sql command and parameter set against the threshold value
+                // in its configuration. If the executed command count is greater then the threshold, it raises a
+                // CounterThresholdException. So here we set the minimum required value to avoid it.
+                configuration.CounterConfiguration.Running.DbCommandExecutionRepetitionThreshold = 26;
```
Also demonstrate how this can be set differently for just a given page. We'll tune these values for each feature/page we test.

Piedone commented on Lombiq.Tests.UI.Samples/Tests/BasicOrchardFeaturesTests.cs:
```diff
@@ -25,7 +25,17 @@ public BasicOrchardFeaturesTests(ITestOutputHelper testOutputHelper)
+            configuration =>
+            {
+                // The UI Testing Toolbox includes DbCommand execution counter, after the end of test it checks the
+                // number of executed commands with the same Sql command and parameter set against the threshold value
```
It's only really the SQL command, regardless of the parameters, that we need to count. A fully repeated query is an issue, but it's a subset of duplicated queries, especially if we want to guard against `SELECT` N+1 queries (where the same query is executed repeatedly, but with different parameters).

Piedone replied:
I'm not sure if this is addressed. Aren't still only duplications with the same query and parameters counted?

dministro replied:
You can configure both. `CounterThresholdConfiguration.DbCommandExecutionThreshold` is used against the count of the same command text with the same parameter value(s), while `CounterThresholdConfiguration.DbCommandTextExecutionThreshold` is used against the count of the same command text.

Piedone replied:
I see, OK then.

Piedone commented on Lombiq.Tests.UI.Samples/Tests/BasicOrchardFeaturesTests.cs:
```diff
@@ -25,7 +25,17 @@ public BasicOrchardFeaturesTests(ITestOutputHelper testOutputHelper)
+            browser,
+            configuration =>
+            {
+                // The UI Testing Toolbox includes DbCommand execution counter, after the end of test it checks the
```
Demonstrating this feature looks more suitable for its own sample test.

Piedone commented on Lombiq.Tests.UI/Exceptions/CounterThresholdException.cs:
```diff
@@ -0,0 +1,64 @@
+    {
+    }
+
+    private static string FormatMessage(
```
Generate exception messages that are easier to understand than:

> DbExecuteCounterKey
> 	SELECT [Document].* FROM [Document] WHERE [Document].[Type] = @Type LIMIT 1
> 		[0]Type = OrchardCore.Settings.SiteSettings, OrchardCore.Settings
> 
> IntegerCounterValue value: 26
> Counter value is greater then DbCommandExecutionRepetitionThreshold, threshold: 22

I as an ordinary developer who didn't set up these counter thresholds but just wanted to change something unrelated but accidentally implemented a `SELECT` N+1 issue, should be able to understand what I did wrong.

Piedone replied:
While the current message is better, it's still cryptic when you first see it.

> SessionProbe, [GET]https://localhost:9341/Admin
Database reader read counter
	Query: SELECT [Document].* FROM [Document] WHERE [Document].[Type] = @Type LIMIT 1
		Parameters: [0]Type = OrchardCore.AdminMenu.Models.AdminMenuList, OrchardCore.AdminMenu
Count: 2
Counter value is greater then SessionThreshold.DbReaderReadThreshold, threshold: 0.

Start the exception with a message that explains, in plain English, what happened.

Piedone replied:
This isn't addressed.

dministro replied:
It was addressed, see:
```
2025-07-16T01:22:23.9882241Z  A counter value has crossed the configured threshold level. Details:
2025-07-16T01:22:23.9882342Z  SessionProbe, [GET]https://localhost:9031/Admin
2025-07-16T01:22:23.9882454Z  Database reader read counter
2025-07-16T01:22:23.9882629Z  	Query: SELECT [Document].* FROM [Document] WHERE [Document].[Type] = @Type LIMIT 1
2025-07-16T01:22:23.9882814Z  		Parameters: [0]Type = OrchardCore.AdminMenu.Models.AdminMenuList, OrchardCore.AdminMenu
2025-07-16T01:22:23.9882866Z  Count: 2
2025-07-16T01:22:23.9883060Z  Counter value is greater then SessionThreshold.DbReaderReadThreshold, threshold: 0.
```

Feel free to add more explanation.

Piedone replied:
Re: https://github.com/Lombiq/UI-Testing-Toolbox/pull/216#issuecomment-3085570127

This is still very technical. "A counter value has crossed the configured threshold level." tells what happened internally. It's like a debug message for the counter infrastructure. But it doesn't help the developer, who'll see their test fail because of this, and who might have not configured this, what actually the problem is.

E.g. something like "The same database query was executed more times than the threshold configuration allows, potentially causing performance issues with excessive database usage. This can indicate a SELECT N+1 problem, or a case where the same query is executed from multiple places in the code instead of using a common cache. See the details below."

And similarly for all kinds of failures.

Piedone commented on Lombiq.Tests.UI/Services/CounterConfiguration.cs:
```diff
@@ -0,0 +1,62 @@
+                        probe,
+                        key,
+                        counterValue,
+                        $"Counter value is greater then {thresholdName}, threshold: {threshold.ToTechnicalString()}");
```
```suggestion
                        $"Counter value is greater then {thresholdName}, threshold: {threshold.ToTechnicalString()}.");
```

Piedone commented on Lombiq.Tests.UI/Services/OrchardCoreHosting/OrchardApplicationFactory.cs:
```diff
@@ -25,16 +27,19 @@ public sealed class OrchardApplicationFactory<TStartup> : WebApplicationFactory<
     private readonly Action<ConfigurationManager, OrchardCoreBuilder> _configureOrchard;
-    private readonly List<IStore> _createdStores = new();
+    private readonly ConcurrentBag<IStore> _createdStores = new();
+    private readonly ICounterDataCollector _counterDataCollector;
```
Keep the order of private fields and ctor assignments the same as the ctor parameters.

Piedone commented on Lombiq.Tests.UI/Services/Counters/Data/ProbedDbConnection.cs:
```diff
@@ -0,0 +1,90 @@
+
+namespace Lombiq.Tests.UI.Services.Counters.Data;
+
+[DesignerCategory("")]
```
Comment why this is necessary.

Piedone commented on Lombiq.Tests.UI/Services/Counters/CounterProbe.cs:
```diff
@@ -0,0 +1,63 @@
+
+namespace Lombiq.Tests.UI.Services.Counters;
+
+public abstract class CounterProbe : CounterProbeBase, IDisposable
```
After a page change, write all counters' values to the TestOutput for debugging.

Piedone reviewed (changes requested):
Impressive!

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Update Lombiq.Tests.UI/Services/CounterDataCollector.cs

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

dministro committed:
Update Lombiq.Tests.UI/Services/Counters/CounterProbeBase.cs

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

dministro committed:
Update Lombiq.Tests.UI/Services/Counters/ICounterValue.cs

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

dministro committed:
Update Lombiq.Tests.UI/Services/Counters/NavigationProbe.cs

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

dministro committed:
Update Lombiq.Tests.UI.Samples/Tests/BasicOrchardFeaturesTests.cs

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

dministro committed:
Update Lombiq.Tests.UI/Services/CounterConfiguration.cs

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

dministro committed:
Merge branch 'issue/OSOE-353' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Adding SessionProbe

dministro committed:
Adding better exclusion and defaults

dministro committed:
Addressing "Add docs to the properties."
Addressing "This needs to be possible to be completely disabled too."
Addressing ""human-readable" everywhere."
Addressing counting on page load not just navigation request
Addressing counting command text with and without parameters
Addressing "Keep the order of private fields and ctor assignments the same as the ctor parameters."
Addressing "Comment why this is necessary."
Addressing "After a page change, write all counters' values to the TestOutput for debugging."

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Fixes after merge

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Fixing analyzer errors

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Fixes after merge

dministro committed:
Fixing possible thread safety violation

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro commented on Lombiq.Tests.UI/Services/UITestExecutor.cs:
```diff
@@ -51,15 +51,16 @@ public static class UITestExecutor
         {
-            lock (_numberOfTestsLimitLock)
-            {
-                _numberOfTestsLimit ??= new SemaphoreSlim(configuration.MaxParallelTests);
-            }
+            _numberOfTestsLimit = new SemaphoreSlim(configuration.MaxParallelTests);
         }
```
[InferSharp ](https://github.com/microsoft/infersharp) reported a thread safety violation here, I can agree with it.
![image](https://user-images.githubusercontent.com/15349881/227495008-2df753c0-f189-4e9d-b8e3-eb5ae261e9f8.png)
Description: Read/Write race. Non-private method `UITestExecutor.ExecuteOrchardCoreTestAsync(...)` reads without synchronization from `Services.UITestExecutor.Lombiq.Tests.UI.Services.UITestExecutor._numberOfTestsLimit`. Potentially races with write in method `UITestExecutor.ExecuteOrchardCoreTestAsync(...)`. Reporting because this access may occur on a background thread.

Piedone replied:
Could you please fix this?

dministro replied:
I fixed it already, just noted why I changed something out of scope.

Piedone replied:
Ah, great! BTW: https://github.com/Lombiq/GitHub-Actions/issues/219

dministro replied:
Agree

dministro committed:
Addressing "S4457: Split this method into two..." analyzer error

dministro committed:
Adding per url threshold configuration support

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Adding per page configuration implementation

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Fixing analyzer violations

dministro committed:
Adding ProbedSqliteConnection

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Improving log and exception message

dministro committed:
Improving logs
Excluding unnecessary counts

dministro committed:
Throwing session probe exception in test context

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Fixing test
Adding docs

dministro committed:
Adding comment

dministro committed:
Fix spelling

Piedone commented on Lombiq.Tests.UI/Services/Counters/Configuration/PhaseCounterConfiguration.cs:
```diff
@@ -0,0 +1,14 @@
+public class PhaseCounterConfiguration : CounterConfiguration
+{
+    /// <summary>
+    /// Gets or sets threshold configuration used under app phase(setup, running) lifetime.
```
```suggestion
    /// Gets or sets threshold configuration used under the app phase (setup, running) lifetime.
```

Piedone commented on Lombiq.Tests.UI.Samples/Tests/BasicOrchardFeaturesTests.cs:
```diff
@@ -25,7 +25,17 @@ public BasicOrchardFeaturesTests(ITestOutputHelper testOutputHelper)
+                // The UI Testing Toolbox includes a DbCommand execution counter. After the end of the test, it checks the
+                // number of executed commands with the same SQL command and parameter set against the threshold value
+                // in its configuration. If the executed command count is greater than the threshold, it raises a
+                // CounterThresholdException. So here we set the minimum required value to avoid it.
```
```suggestion
                // The UI Testing Toolbox includes a DbCommand execution counter to check for duplicated SQL queries..
                // After the end of the test, it checks the number of executed commands with the same SQL command text
                // and parameter set against the threshold value in its configuration. If the executed command count is
                // greater than the threshold, it raises a CounterThresholdException. So here we set the minimum
                // required value to avoid it.
```

Piedone commented on Lombiq.Tests.UI.Samples/Tests/DuplicatedSqlQueryDetectorTests.cs:
```diff
@@ -0,0 +1,60 @@
+namespace Lombiq.Tests.UI.Samples.Tests;
+
+// Some times you may want to detect duplicated SQL queries. This can be useful if you want to make sure that your code
+// does not execute the same query multiple times.
```
```suggestion
// does not execute the same query multiple times, wasting time and computing resources.
```

Piedone commented on Lombiq.Tests.UI.Samples/Tests/DuplicatedSqlQueryDetectorTests.cs:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    // This test will fail because the app will read the the same command result more times then the configured threshold
+    // during the Admin page rendering.
```
```suggestion
    // This test will fail because the app will read the same command result more times than the configured threshold
    // during the Admin page rendering.
```

Piedone commented on Lombiq.Tests.UI.Samples/Tests/DuplicatedSqlQueryDetectorTests.cs:
```diff
@@ -0,0 +1,60 @@
+                browser,
+                ConfigureAsync));
+
+    // We configure the test to throw an exception if a certain counter threshold is exceeded, but only in case of Admin page.
```
```suggestion
    // We configure the test to throw an exception if a certain counter threshold is exceeded, but only in case of Admin
    // pages.
```

Piedone commented on Lombiq.Tests.UI.Samples/Tests/DuplicatedSqlQueryDetectorTests.cs:
```diff
@@ -0,0 +1,60 @@
+        // Let's enable and configure the counter threshold for ORM sessions.
+        adminCounterConfiguration.SessionThreshold.Disable = false;
+        adminCounterConfiguration.SessionThreshold.DbReaderReadThreshold = 0;
+        // Apply the configuration to the Admin page only.
```
```suggestion
        // Apply the configuration to the Admin pages only.
```

Piedone commented on Lombiq.Tests.UI/Services/Counters/Configuration/CounterThresholdConfiguration.cs:
```diff
@@ -0,0 +1,29 @@
+    /// <summary>
+    /// Gets or sets a value indicating whether the current threshold configuration and checking is disabled.
+    /// </summary>
+    public bool Disable { get; set; } = true;
```
Use `IsEnabled` instead. This is more commonly used in APIs. Also, why is this property necessary in the first place? When would you add a configuration but have it disabled?

Piedone replied:
You didn't answer here. Why do we have both this and `CounterConfigurations.IsEnabled`?

dministro replied:
We can disable the assertion for the given "lifecycle". (Navigation, Session, PageLoad)

Piedone replied:
I see, OK then.

Piedone commented on Lombiq.Tests.UI/Services/Counters/Configuration/CounterThresholdConfiguration.cs:
```diff
@@ -0,0 +1,29 @@
+    /// <see cref="System.Data.Common.DbCommand.CommandText"/> and <see cref="System.Data.Common.DbCommand.Parameters"/>
+    /// for counting.
+    /// </summary>
+    public int DbCommandExecutionThreshold { get; set; } = 11;
```
What are the magic 11s?

Piedone replied:
```suggestion
    public int DbCommandIncludingParametersExecutionCountThreshold { get; set; } = 11;
```

Piedone commented on Lombiq.Tests.UI.Samples/Tests/DuplicatedSqlQueryDetectorTests.cs:
```diff
@@ -0,0 +1,60 @@
+    // This test will pass because not the Admin page was loaded.
+    [Theory, Chrome]
+    public Task DbReaderReadDuringRunningPhaseShouldNotThrow(Browser browser) =>
+        Should.NotThrowAsync(() =>
```
`Should.NotThrowAsync()` is not needed.

Piedone commented on Lombiq.Tests.UI.Samples/Tests/DuplicatedSqlQueryDetectorTests.cs:
```diff
@@ -0,0 +1,60 @@
+
+    // This test will pass because not the Admin page was loaded.
+    [Theory, Chrome]
+    public Task DbReaderReadDuringRunningPhaseShouldNotThrow(Browser browser) =>
```
Demonstrate adding a config for this page that has a threshold that matches the page.

Piedone replied:
```suggestion
    public Task PageWithoutDuplicatedSqlQueriesShouldPass(Browser browser) =>
```

Piedone commented on Lombiq.Tests.UI.Samples/Tests/DuplicatedSqlQueryDetectorTests.cs:
```diff
@@ -0,0 +1,60 @@
+    // This test will fail because the app will read the the same command result more times then the configured threshold
+    // during the Admin page rendering.
+    [Theory, Chrome]
+    public Task DbReaderReadDuringRunningPhaseShouldThrow(Browser browser) =>
```
```suggestion
    public Task PageWithTooManyDuplicatedSqlQueriesShouldThrow(Browser browser) =>
```

Piedone commented on Lombiq.Tests.UI.Samples/Tests/BasicOrchardFeaturesTests.cs:
```diff
@@ -32,7 +32,7 @@ public BasicOrchardFeaturesTests(ITestOutputHelper testOutputHelper)
                 // in its configuration. If the executed command count is greater than the threshold, it raises a
                 // CounterThresholdException. So here we set the minimum required value to avoid it.
-                configuration.CounterConfiguration.Running.DbCommandExecutionRepetitionThreshold = 26;
+                configuration.CounterConfiguration.Running.PhaseThreshold.DbCommandExecutionThreshold = 26;
```
There are 26 duplicated queries during this test?

Piedone commented on Lombiq.Tests.UI/Services/Counters/IRelativeUrlConfigurationKey.cs:
```diff
@@ -0,0 +1,19 @@
+/// <summary>
+/// Represents a relative URL configuration key in <see cref="Configuration.RunningPhaseCounterConfiguration"/>.
+/// </summary>
+public interface IRelativeUrlConfigurationKey : ICounterConfigurationKey
```
This would be more straightforward with regex matching.

Piedone replied:
This isn't addressed.

dministro replied:
I feel it is a bit overengineering. We can do it later under a different PR if it really necessary.

Piedone replied:
Why overengineering? We're matching a URL against a substring, with an additional Boolean switch (`ExactMatch`). Matching it against a regex is a very small next step, but provides ultimate flexibility, while removing the switch.

I've written this almost two years ago, we won't defer this anymore.

Piedone commented on Lombiq.Tests.UI/Services/Counters/SessionProbe.cs:
```diff
@@ -0,0 +1,71 @@
+    async ValueTask IAsyncDisposable.DisposeAsync()
+    {
+        await _session.DisposeAsync();
+        // Should be at the end because, the Session implementation calls CommitOrRollbackTransactionAsync in DisposeAsync
+        // and we should count the executed db commands in it.
```
```suggestion
        // Should be at the end because the Session implementation calls CommitOrRollbackTransactionAsync in
        // DisposeAsync and we should count the executed DB commands in it.
```

Piedone commented on Lombiq.Tests.UI.Samples/Tests/BasicOrchardFeaturesTests.cs:
```diff
@@ -25,7 +25,17 @@ public BasicOrchardFeaturesTests(ITestOutputHelper testOutputHelper)
+                // number of executed commands with the same SQL command and parameter set against the threshold value
+                // in its configuration. If the executed command count is greater than the threshold, it raises a
+                // CounterThresholdException. So here we set the minimum required value to avoid it.
+                configuration.CounterConfiguration.Running.PhaseThreshold.DbCommandExecutionThreshold = 26;
```
Why is this configuration needed in this test?

Piedone commented on Lombiq.Tests.UI/Services/Counters/Configuration/CounterConfiguration.cs:
```diff
@@ -0,0 +1,86 @@
+
+    public static IEnumerable<ICounterKey> DefaultExcludeList { get; } = new List<ICounterKey>
+    {
+        new DbCommandExecuteCounterKey(
+            WorkflowTypeStartActivitiesQuery,
+            new List<KeyValuePair<string, object>>
+            {
+                new("p0", "ContentCreatedEvent"),
+                new("p1", value: true),
+            }),
+        new DbCommandExecuteCounterKey(
+            WorkflowTypeStartActivitiesQuery,
+            new List<KeyValuePair<string, object>>
+            {
+                new("p0", "ContentPublishedEvent"),
+                new("p1", value: true),
+            }),
+        new DbCommandExecuteCounterKey(
+            WorkflowTypeStartActivitiesQuery,
+            new List<KeyValuePair<string, object>>
+            {
+                new("p0", "ContentUpdatedEvent"),
+                new("p1", value: true),
+            }),
```
Why are these excluded?

dministro replied:
Happening many times during the setup, so I found them useless. Used only in samples, so I can move there, and double check it.

Piedone replied:
That would be better.

Piedone commented on Lombiq.Tests.UI/Services/Counters/ICounterDataCollector.cs:
```diff
@@ -26,4 +28,9 @@ public interface ICounterDataCollector : ICounterProbe
     void AssertCounter();
+
+    /// <summary>
+    /// Postpones exception thrown by a counter in case when the exception was thrown out of the test context.
```
```suggestion
    /// Postpones exception thrown by a counter when the exception was thrown from the test context.
```

Piedone commented on Lombiq.Tests.UI/Services/Counters/ICounterKey.cs:
```diff
@@ -8,8 +9,13 @@ namespace Lombiq.Tests.UI.Services.Counters;
+    /// <summary>
+    /// Dumps the key content to a human-readable format.
+    /// </summary>
+    /// <returns>A human-readable representation of instance.</returns>
```
```suggestion
    /// <returns>A human-readable representation of the instance.</returns>
```

Piedone commented on Lombiq.Tests.UI/Services/Counters/IRelativeUrlConfigurationKey.cs:
```diff
@@ -0,0 +1,19 @@
+    Uri Url { get; }
+
+    /// <summary>
+    /// Gets a value indicating whether the URL should be matched exactly or substring match enough.
```
```suggestion
    /// Gets a value indicating whether the URL should be matched exactly or substring match is enough.
```

Piedone commented on Lombiq.Tests.UI/Services/Counters/Middlewares/PageLoadProbeMiddleware.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace Lombiq.Tests.UI.Services.Counters.Middlewares;
+
+public class PageLoadProbeMiddleware
```
Why is this and `NavigationProbe` both necessary?

dministro replied:
The `NavigationProbe` is for the navigation requested by the test code e.g. `UITestContext.GoToHomePageAsync(..)` while the `PageLoadProbe` is for the requests from the web browser e.g. image requests, API calls, etc. The `NavigationProbe` captures image requests while the page loading, but not async calls.

Piedone replied:
Wouldn't `PageLoadProbe` alone enough? Since all navigations requested by the test will in the end result in a request.

BTW I'd rather call this `RequestLoadProbleMiddleware` then.

dministro replied:
Yes, it results at least one request but the kex difference is not just the operations counted during the page generation but all operations (eg image - included into the page - loaded from the DB). So it counts SQL operations for multiple request connected for the given page load.

Piedone replied:
kex difference?

![image](https://github.com/user-attachments/assets/cb08213f-4178-4d16-b237-b79275b49b5f)

But OK then :).

Piedone commented on Lombiq.Tests.UI/Services/ProbedConnectionFactory.cs:
```diff
@@ -19,6 +20,14 @@ public ProbedConnectionFactory(IConnectionFactory connectionFactory, ICounterDat
+    {
+        var connection = _connectionFactory.CreateConnection();
+
+        // This consition and the ProbedSqliteConnection can be removed once
+        // https://github.com/OrchardCMS/OrchardCore/issues/14217 get fixed.
```
```suggestion
        // This condition and the ProbedSqliteConnection can be removed once
        // https://github.com/OrchardCMS/OrchardCore/issues/14217 gets fixed.
```

dministro committed:
Apply suggestions from code review

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

Piedone commented on Lombiq.Tests.UI/Services/Counters/Configuration/CounterThresholdConfiguration.cs:
```diff
@@ -0,0 +1,29 @@
+    /// Gets or sets the threshold of executed <see cref="System.Data.Common.DbCommand"/>s. Uses
+    /// <see cref="System.Data.Common.DbCommand.CommandText"/> for counting.
+    /// </summary>
+    public int DbCommandTextExecutionThreshold { get; set; } = 11;
```
```suggestion
    public int DbCommandExcludingParametersExecutionThreshold { get; set; } = 11;
```

Piedone commented on Lombiq.Tests.UI/Services/Counters/Configuration/CounterThresholdConfiguration.cs:
```diff
@@ -0,0 +1,29 @@
+    /// </summary>
+    public bool Disable { get; set; } = true;
+
+    /// <summary>
+    /// Gets or sets the threshold of executed <see cref="System.Data.Common.DbCommand"/>s. Uses
+    /// <see cref="System.Data.Common.DbCommand.CommandText"/> and <see cref="System.Data.Common.DbCommand.Parameters"/>
+    /// for counting.
+    /// </summary>
```
```suggestion
    /// <summary>
    /// Gets or sets the threshold for the count of <see cref="System.Data.Common.DbCommand"/> executions, with the
    /// query only counted as a duplicate if both its text (<see cref="System.Data.Common.DbCommand.CommandText"/>) and
    /// parameters (<see cref="System.Data.Common.DbCommand.Parameters"/>) match. See
    /// <see cref="DbCommandTextExecutionThreshold"/> for counting using only the command text.
    /// </summary>
```

Piedone commented on Lombiq.Tests.UI/Services/Counters/Configuration/CounterThresholdConfiguration.cs:
```diff
@@ -0,0 +1,29 @@
+    /// </summary>
+    public int DbCommandExecutionThreshold { get; set; } = 11;
+
+    /// <summary>
+    /// Gets or sets the threshold of executed <see cref="System.Data.Common.DbCommand"/>s. Uses
+    /// <see cref="System.Data.Common.DbCommand.CommandText"/> for counting.
+    /// </summary>
```
```suggestion
    /// <summary>
    /// Gets or sets the threshold for the count of <see cref="System.Data.Common.DbCommand"/> executions, with the
    /// query counted as a duplicate if its text (<see cref="System.Data.Common.DbCommand.CommandText"/>) matches.
    /// Parameters (<see cref="System.Data.Common.DbCommand.Parameters"/>) are not taken into account. See
    /// <see cref="DbCommandExecutionThreshold"/> for counting using also the parameters.
    /// </summary>
```

dministro committed:
Apply suggestions from code review

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

Piedone commented:
@coderabbitai review

I'm testing the AI code review tool [CodeRabbit](https://coderabbit.ai/) here. Please bear with me. And if it comments some useful feedback, please address it.

Piedone commented:
@coderabbitai review

Sorry about that.

Piedone commented:
These comments seem useful, please check them out. You can also chat with the bot, see the tips under https://github.com/Lombiq/UI-Testing-Toolbox/pull/216#issuecomment-1872376517.

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Fixes after merge

dministro committed:
Fixing analyzer violations

dministro committed:
Fixing analyzer violations again

Piedone commented on Lombiq.Tests.UI/Services/Counters/Configuration/CounterConfiguration.cs:
```diff
@@ -0,0 +1,86 @@
+    /// <see cref="UI.Extensions.NavigationUITestContextExtensions.GoToAbsoluteUrlAsync(UITestContext, Uri, bool)"/>.
+    /// See: <see cref="NavigationProbe"/>.
+    /// </summary>
+    public CounterThresholdConfiguration NavigationThreshold { get; set; } = new CounterThresholdConfiguration
+    {
+        DbCommandIncludingParametersExecutionCountThreshold = 11,
+        DbCommandExcludingParametersExecutionThreshold = 22,
+        DbReaderReadThreshold = 11,
+    };
+
+    /// <summary>
+    /// Gets or sets threshold configuration used per <see cref="YesSql.ISession"/> lifetime. See:
+    /// <see cref="SessionProbe"/>.
+    /// </summary>
+    public CounterThresholdConfiguration SessionThreshold { get; set; } = new CounterThresholdConfiguration
+    {
+        DbCommandIncludingParametersExecutionCountThreshold = 22,
+        DbCommandExcludingParametersExecutionThreshold = 44,
+        DbReaderReadThreshold = 11,
+    };
+
+    /// <summary>
+    /// Gets or sets threshold configuration used per page load. See: <see cref="PageLoadProbe"/>.
+    /// </summary>
+    public CounterThresholdConfiguration PageLoadThreshold { get; set; } = new CounterThresholdConfiguration
+    {
+        DbCommandIncludingParametersExecutionCountThreshold = 22,
+        DbCommandExcludingParametersExecutionThreshold = 44,
+        DbReaderReadThreshold = 11,
+    };
```
Please add a piece of docs contrasting these three configs: when would you want to use each?

Piedone replied:
This isn't addressed.

dministro committed:
Adding custom type for storing db command parameters in configuration instead of KeyValuePair<>
Setting default configuration more generic
Moving OSOCE default configuration to OrchardCoreUITestExecutorConfiguration

dministro committed:
Allowing to enable/disable counter subsystem by configuration
Fine-tuning thresholds in tests

dministro committed:
Adding a bit more information to CounterThresholdException message

dministro committed:
Removing unnecessary Should.NotThrowAsync()

dministro committed:
Adding test to demonstrate the scenario when the counter thresholds are exactly matching with the counter values captured during navigation.

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Removing unnecessary configuration

dministro committed:
Renaming `PageLoadProbe` -> `RequestProbe` and `PageLoadProbeMiddleware` -> `RequestProbeMiddleware`

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Adding more comments
Adding more sample tests

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Spelling

dministro committed:
Fixing analyzer warnings

dministro committed:
Fix wording

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Updating CompatibilitySuppressions.xml

dministro committed:
Fixing analyzer violations

dministro committed:
Implementing IDisposable properly

dministro committed:
Comment

dministro committed:
Fixing analyzer violation

dministro committed:
Fixing possible NRE

dministro committed:
Looping DuplicatedSqlQueryDetectorTests.cs

dministro committed:
Adding more docs

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Introducing review comments

dministro committed:
Fixing analyzer violation

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Fixes after merge

dministro committed:
Removing leftover

dministro committed:
Suppressing breaking changes

dministro committed:
Suppressing breaking changes

dministro committed:
Moving sample configuration to samples

dministro committed:
Merge branch 'dev' of https://github.com/Lombiq/UI-Testing-Toolbox into issue/OSOE-353

dministro committed:
Upgrading OC modules to 2.1.7
Implementing missing methods

Piedone committed:
Merge branch 'dev' into issue/OSOE-353

Piedone commented on Lombiq.Tests.UI.Samples/Tests/DuplicatedSqlQueryDetectorTests.cs:
These tests log failures much like actually failing tests, causing error outputs and tests dumps, see e.g. https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/actions/runs/16277477256. Instead of having failing `ExecuteTestAfterSetupAsync`s, do what `ErrorHandlingTests` does and catch the assertion exceptions.

Piedone commented on Lombiq.Tests.UI/Docs/CounterInfrastructure.md:
```diff
@@ -0,0 +1,41 @@
+
+- **Navigation:** Captures operations performed during page load, including SQL queries executed for page generation and content retrieval (e.g., images from the database). The configuration is stored in `NavigationThreshold`.
+- **Page Load:** Captures operations performed strictly during page load. The configuration is stored in `PageLoadThreshold`.
+- **Session:** Captures operations executed while a given `ISession` instance is alive (e.g., a session used by a background task). The configuration is stored in `SessionThreshold`.
```
Instead of explaining the properties here, merge this with the docs on the properties, so ultimately, the properties have the complete docs. Then, here just offer a quick shorter rundown of 3-3 bullet points on what the counters count, and when they are asserted, referring people to read the docs on the classes for in-depth info.

dministro replied:
I moved the CounterThresholdConfiguration property documentation to the code, but the lifecycle description is better here.

Piedone commented on Lombiq.Tests.UI.Samples/Tests/DuplicatedSqlQueryDetectorTests.cs:
```diff
@@ -0,0 +1,141 @@
+    {
+    }
+
+    // This test will fail because the app will read the same command result more times than the configured threshold
+    // during the Admin page rendering.
```
This is a technical description of what happens in the counter logic. This is good, but add an explanation of why this happens in higher-level terms. I.e., there's a query on the OC admin that's duplicated or something, mention that.

Same for the other two ShouldThrow samples.

dministro replied:
Sorry, but the current comment is what you asked. What higher level do you think?

Piedone replied:
What causes this, what's the command, what's the admin actually doing to read more?

Piedone commented on Lombiq.Tests.UI.Samples/Tests/DuplicatedSqlQueryDetectorTests.cs:
```diff
@@ -0,0 +1,141 @@
+// does not execute the same query multiple times, wasting time and computing resources.
+public class DuplicatedSqlQueryDetectorTests : UITestBase
+{
+    private const string WorkflowTypeStartActivitiesQuery =
+        "SELECT DISTINCT [Document].* FROM [Document] INNER JOIN [WorkflowTypeStartActivitiesIndex]"
+        + " AS [WorkflowTypeStartActivitiesIndex_a1]"
+        + " ON [WorkflowTypeStartActivitiesIndex_a1].[DocumentId] = [Document].[Id]"
+        + " WHERE (([WorkflowTypeStartActivitiesIndex_a1].[StartActivityName] = @p0)"
+        + " and ([WorkflowTypeStartActivitiesIndex_a1].[IsEnabled] = @p1))";
+
+    private static readonly IEnumerable<ICounterKey> CounterExcludeList =
+    [
+        new DbCommandExecuteCounterKey(
+            WorkflowTypeStartActivitiesQuery,
+            new("p0", "ContentCreatedEvent"),
+            new("p1", value: true)),
+        new DbCommandExecuteCounterKey(
+            WorkflowTypeStartActivitiesQuery,
+            new("p0", "ContentPublishedEvent"),
+            new("p1", value: true)),
+        new DbCommandExecuteCounterKey(
+            WorkflowTypeStartActivitiesQuery,
+            new("p0", "ContentUpdatedEvent"),
+            new("p1", value: true)),
+    ];
```
Add a comment about what this is.

dministro replied:
Comment added to where the configuration is happening.

Piedone commented on Lombiq.Tests.UI.Samples/Tests/DuplicatedSqlQueryDetectorTests.cs:
```diff
@@ -0,0 +1,141 @@
+namespace Lombiq.Tests.UI.Samples.Tests;
+
+// Some times you may want to detect duplicated SQL queries. This can be useful if you want to make sure that your code
+// does not execute the same query multiple times, wasting time and computing resources.
```
I added a note here, but explain how you'd start using this in your app. E.g., do you just start using it all of your tests with the defaults? Or only on the frontend?

dministro replied:
Sorry, but I don't understand what you are requesting. This feature is connected to the persistent layer, so how the frontend is coming to the picture?

Piedone replied:
By frontend I mean the public-facing part of the site, i.e. not the admin.

Piedone commented on Lombiq.Tests.UI/Extensions/NavigationUITestContextExtensions.cs:
```diff
@@ -51,9 +52,12 @@ public static Task GoToAbsoluteUrlAsync(this UITestContext context, Uri absolute
-                // Navigation can sometimes not happen on the first try.
-                await context.DoWithRetriesUntilNavigationHasOccurredOrFailAsync(
-                    () => context.Driver.Navigate().GoToUrlAsync(absoluteUri));
+                using (new NavigationProbe(context.CounterDataCollector, absoluteUri))
```
Do we always need a `NavigationProbe` instantiation here, even if the test suite doesn't use the counters at all? Can't this be done with an optional `NavigationEventHandler`?

Piedone commented on Lombiq.Tests.UI/Services/Counters/Configuration/PhaseCounterConfiguration.cs:
```diff
@@ -0,0 +1,14 @@
+    /// </summary>
+    public CounterThresholdConfiguration PhaseThreshold { get; set; } = new CounterThresholdConfiguration
+    {
+        DbCommandIncludingParametersExecutionCountThreshold = 22,
+        DbCommandExcludingParametersExecutionThreshold = 44,
+        DbReaderReadThreshold = 11,
```
Where do these numbers come from?

Piedone commented on Lombiq.Tests.UI/Services/OrchardCoreHosting/OrchardApplicationFactory.cs:
```diff
@@ -103,8 +115,13 @@ .ImplementationInstance as ConfigurationManager
             {
                 AddFakeStore(builderServices);
                 AddFakeViewCompilerProvider(builderServices);
+                AddSessionProbe(builderServices);
```
Anything related to counters should only be done here if they're actually enabled. So, the session probe and middleware shouldn't be added if it's disabled, nor `ProbedConnectionFactory`.

Piedone commented on Lombiq.Tests.UI/Services/OrchardCoreUITestExecutorConfiguration.cs:
```diff
@@ -211,4 +223,78 @@ public CancellationToken TestCancellationToken
+                        settings.Settings.DbCommandIncludingParametersExecutionCountThreshold);
+                    AssertIntegerCounterValue<DbCommandTextExecuteCounterKey>(
+                        probe,
+                        counterConfiguration.ExcludeFilter ?? (key => false),
+                        $"{settings.Name}.{nameof(settings.Settings.DbCommandExcludingParametersExecutionThreshold)}",
+                        settings.Settings.DbCommandExcludingParametersExecutionThreshold);
+                    AssertIntegerCounterValue<DbReaderReadCounterKey>(
+                        probe,
+                        counterConfiguration.ExcludeFilter ?? (key => false),
+                        $"{settings.Name}.{nameof(settings.Settings.DbReaderReadThreshold)}",
+                        settings.Settings.DbReaderReadThreshold);
+                }
+                catch (CounterThresholdException exception) when (probe is IOutOfTestContextCounterProbe)
+                {
+                    collector.PostponeCounterException(exception);
+                }
+            }
+        };
+
+    public static void AssertIntegerCounterValue<TKey>(
+        ICounterProbe probe,
+        Func<ICounterKey, bool> excludeFilter,
+        string thresholdName,
+        int threshold)
+        where TKey : ICounterKey =>
+        probe.Counters.Keys
+            .OfType<TKey>()
+            .Where(key => !excludeFilter(key))
+            .ForEach(key =>
+            {
+                if (probe.Counters[key] is IntegerCounterValue counterValue
+                    && counterValue.Value > threshold)
+                {
+                    throw new CounterThresholdException(
+                        probe,
+                        key,
+                        counterValue,
+                        $"Counter value is greater then {thresholdName}, threshold: {threshold.ToTechnicalString()}.");
+                }
+            });
```
I don't think these need to be in this class.

Piedone commented on Lombiq.Tests.UI/Services/UITestExecutionSession.cs:
```diff
@@ -119,6 +120,10 @@ public async Task<bool> ExecuteAsync(int retryCount, string dumpRootPath)
             _context.TestDumpContainer.Clear();
             testDumpContainer = _context.TestDumpContainer;
 
+            BeginDataCollection(
```
Only run this and `EndAssertDataCollection` when the counters are enabled.

Piedone reviewed (changes requested):
Be sure to address the older comments too.

Piedone committed:
Grammar/clarifying docs, code styling

dministro commented:
I can't answer under your comments:
https://github.com/Lombiq/UI-Testing-Toolbox/pull/216#discussion_r1370964780
https://github.com/Lombiq/UI-Testing-Toolbox/pull/216#discussion_r2208939079

It was addressed, see:
```
2025-07-16T01:22:23.9882241Z  A counter value has crossed the configured threshold level. Details:
2025-07-16T01:22:23.9882342Z  SessionProbe, [GET]https://localhost:9031/Admin
2025-07-16T01:22:23.9882454Z  Database reader read counter
2025-07-16T01:22:23.9882629Z  	Query: SELECT [Document].* FROM [Document] WHERE [Document].[Type] = @Type LIMIT 1
2025-07-16T01:22:23.9882814Z  		Parameters: [0]Type = OrchardCore.AdminMenu.Models.AdminMenuList, OrchardCore.AdminMenu
2025-07-16T01:22:23.9882866Z  Count: 2
2025-07-16T01:22:23.9883060Z  Counter value is greater then SessionThreshold.DbReaderReadThreshold, threshold: 0.
```

Feel free to add more explanation.

Piedone commented:
You can reply there, exactly under the link, so I'll also follow up there.

Piedone commented:
Somebody else will do this.
